### ISSUE TITLE
deprecate baseURL

### ISSUE THREAD

**Issue by guybedford** [systemjs/systemjs#290]
I believe this is going out with the move to module names as URIs.

@caridy it would be good to get confirmation on this?

**Comment by matthewp** [comment]
Without a baseURL (or a wildcard path) you can't reuse a config file. The paths become worthless, they have to be redefined for each page.

**Comment by guybedford** [comment]
Ok, agreed we need to keep `baseURL` It is critically necessary for bundling, because the only way to bundle in a URL module scheme is either as absolute URIs or relative to some baseURL. Absolute URI bundles are not portable so to make them portable we need a baseURL the names are relative to.

**Comment by guybedford** [comment]
Not happening.

**Comment by caridy** [comment]
@guybedford I will keep this in mind, we can revisit this later this month.

**Comment by guybedford** [comment]
@caridy sure that makes sense, but as long as it can be implemented in an extension (which it currently can be). As discussed here, we do need this feature after all.

**Comment by matthewp** [comment]
@caridy I'm not sure how it's possible to reuse a config that contains `paths` without baseURL. Reusing a config is *very* important in large applications where you have more than a single index.html. You might have separate test/demo pages for each component you are working on and need to use the same configuration file for each of these.

**Comment by caridy** [comment]
@matthewp this is not different from scripts, and even links in a page. The same principle should apply, including `<base href="/path/to/base/">`. My main concern of introducing a configuration is that script/module tags will not, or should not, rely on that for obvious reasons, which means we will have a dual behavior, and that will be undesired.

**Comment by matthewp** [comment]
Geez, I forgot all about `base`. I remember it having a lot of gotchas in the past but maybe those have been cleared up. Can you put a relative path in the base href (relative to the current document)?  If so that might work. Still have to consider non-browser environments though.

**Comment by guybedford** [comment]
Yeah would one really want the script base and the `<a href="some/path">` bases to match though? This is creating coupled concerns between navigation and script loading.

**Comment by caridy** [comment]
@matthewp yes

**Comment by guybedford** [comment]
I just wanted to come back to this idea of deprecating an explicit baseURL for the loader. The baseURL "under the hood" can be the `base` or `window.location.origin`.

The problem we hit previously was that bundles want to be able to be baseURL-relative and control that baseURL.

But surely we just use absolute URLs for bundles then?

```
System.register('/first/module.js', ...)
```

That is, our bundles still go through the `locate` phase, but not the `normalize` phase (these phases will be simulated within the new resolve hook for conceptual simplicity).

So @matthewp do you have a strong argument that bundle names need more than absolute normalization rules? Note that SystemJS will probably still expose a locate hook so that custom location functions could be written allowing custom concepts of `bundleBase` and bundle portability.

In short, I'm still keen to deprecate baseURL in the loader, and just make the root the natural reference point.

**Comment by matthewp** [comment]
I'm not following your question, maybe it's related to how System.register bundles works (which we don't use)?  I would hope that the bundle extension still normalizes (or whatever that becomes) the bundle names though, I can't think of a reason why not?  I'm probably misunderstanding you though.

**Comment by guybedford** [comment]
Regardless of the exact process, the question is simply if there are scenarios where an explicit baseURL not matching the window.location might be needed, and I'm still not convinced that there are.

**Comment by matthewp** [comment]
Yes, I don't think so.  I haven't used `<base>` in recent years so if it's possible to do `<base href="../../">` then I don't see a need for a SystemJS baseURL.

**Comment by MajorBreakfast** [comment]
I think `baseURL` should go away. Or at least I don't see me ever using it.

BTW in ember apps we extensively use the base tag because the server responds to `example.com` and `example.com/posts/10` with the same `index.html`. The ember app will look differently for both URLs of course - but that's all in the client-side code. The `index.html` always contains a base tag with a domain relative URL, e.g. "/", so that all resources load.

**Comment by theefer** [comment]
I'd stay away from relying on `<base href="...">` as other framework (e.g. Angular) seem to want to (ab)use it for their router and this will likely creating mixing of concerns that will be hard to reconciliate across the various bits that would read this tag.

In our app, we need to override the `baseURL` to point to `/assets`, which is the path under which our server exposes the assets from its `public` local folder. If there were no `baseURL`, how would we be expected to manage this in the configuration?

**Comment by MajorBreakfast** [comment]
> I'd stay away from relying on <base href="..."> as other framework (e.g. Angular) seem to want to (ab)use it for their router

@theefer What do you mean? I think the base tag is the optimal primitive to define what place you consider the app root in relation to the domain. I am not aware of any downsides.

**Comment by probins** [comment]
> if there are scenarios where an explicit baseURL not matching the window.location might be needed

when you are loading from the cloud, not from the server that the base html sits on. My 'app' consists of a small html file that people can copy to Google Drive, Dropbox or similar, and which pulls in the code from jspm:github. I have a testing version which pulls in from localhost, and the only real difference is the baseURL. I might be able to use <base> instead; I have used this in the past and seem to remember coming across issues, but it's so long ago I can't really remember what the issues were.

**Comment by matthewp** [comment]
Actually I think baseURL is essential in Node.  You can't reasonably expect to be able to change `process.chdir()` because other parts of the program (other libraries, etc) will not be expecting it.

**Comment by theefer** [comment]
@MajorBreakfast For Angular, the base URL is essentially the root for the routing. For SystemJS, the base URL is the root for resolving module paths. What if the routing root (say `/`) is not the same as the module path root (say `/assets` or even `https://cdn.example.io/app-files`)?

@guybedford I know we discussed this in previous issues and PRs, but how else than using `baseURL` is one supposed to setup the *server* path where assets are exposed under? Can't really hack the `paths` in the config as they would then no longer match local file paths on disk. I might be missing something, but this seems like a strong need for `baseURL`.

**Comment by guilhermeaiolfi** [comment]
I'm using ```baseURL``` to define the path for production/development/etc based on the 'environment' that my server-side framework is using. That way I keep the same config.js regardless. So it's essential for me or am I doing that wrong?

**Comment by guybedford** [comment]
The idea was to try and converge on absolute URLs (like `/path/to/module.js`) and share that base always, but I guess it is a little too restricting to always expect.

**Comment by theefer** [comment]
For us the problem is that the absolute path of the URL on the server may need to differ from the local path. The common path fragment can indeed be set in the `config.js` `paths`, but the server prefix has to be set using `System.baseURL` whereas the local filesystem prefix is set using `directories.baseURL` in `package.json`.

For example consider:

```
// config.js

System.config({
  "paths": {
    "app/*": "lib/*.js",
    "github:*": "jspm_packages/github/*.js",
    "npm:*": "jspm_packages/npm/*.js"
  }
});


// SERVER
https://example.com/index.html (or https://example.com/somewhere/else/index.html)

<script src="/assets/config.js"></script>
<script>
System.baseURL = '/assets';
System.import('app/main');
</script>

loads:
https://example.com/assets/config.js
https://example.com/assets/lib/main.js


// package.json

{
  "jspm": {
    "directories": {
      "baseURL": "public"
    },
   ...
}

given:
dir/package.json
dir/public/config.js
dir/public/lib/main.js
```

In other words, for paths to resolve correctly on both the server and client, they need different `baseURL`s. I think that's a fairly common pattern really...

**Comment by MajorBreakfast** [comment]
@theefer As a side note: It's possible to dynamically set the base tag just like baseURL:
```HTML
<head>
  <script>
    if (...) { document.write('<base href="...">') }
  </script>
</head>
```
I've done this in the past. It works as expected.

**Comment by theefer** [comment]
@MajorBreakfast But in my example above, it'd have to be set to `/assets` for SystemJS (root of assets paths) and to `/` for AngularJS (navigation). Which one is it?

**Comment by guybedford** [comment]
Thanks everyone for letting me rehash this. It sounds like we will stick with it for practicality.

**Comment by MajorBreakfast** [comment]
@theefer The trick is to avoid that scenario in the first place. For example:
```HTML
<link rel="stylesheet" href="assets/style.css">
```
```JS
System.import('assets/style.css!')
```
It's better if these two access the same file to avoid unnecessary confusion. The `baseURL` then only needs to be used in environments like node oder web workers which have no base tag.

**Comment by guybedford** [comment]
Still going in circles here, but I do still think there is merit to this.

Proposal:

* baseURL is supported only when running builds
* in the browser use path / package configuration to have default resolution use custom subpaths

**Comment by guybedford** [comment]
The basic idea is that `System.import('./x')` should always follow normal URL rules.

**Comment by guybedford** [comment]
Note it would still be possible to have baseURL-style functionality via paths:

```javascript
System.paths['*'] = '/custom/path/*';

System.import('app');
```

The only thing that changes is that this would no longer apply to `System.import('./app')`.

This new resolution algorithm has been a beast, but I think we've finally got something sensible that aligns with the spec as well, and deprecating baseURL (in the browser) is a key part of that.

**Comment by guybedford** [comment]
Merged into master.

**Comment by caridy** [comment]
@guybedford +1, baseURL-style functionality via paths is what we were pushing for during our previous conversations. The owner of the app is in control of the paths and the base, which means they can do whatever they want with it.

**Comment by matthewp** [comment]
@caridy That's not true in Node though, so I don't know what the solution should be there.

**Comment by caridy** [comment]
@matthewp I will like to hear more thoughts about that. I don't see any actual difference, considering that top level definitions are in control in nodejs, which means you can always compute and set the paths at the app level, or at the dependent module level.

**Comment by guybedford** [comment]
@caridy baseURL is still in for now (despite my trying hard to remove it). I'm still on the fence actually about deprecating baseURL.

Basically we can have users do:

```javascript
System.config({
  paths: {
    '*': '/app/*'
  }
});
```

Or we can have users do:

```javascript
System.config({
  baseURL: '/app/'
});
```

So I'd only want to deprecate baseURL when we can be sure of where the spec stands on wildcard sites so we can at least provide a "stable" alternative. Having to be very careful with the upgrade paths these days.

**Comment by alvinsight** [comment]
Hey guys, 
Where are we for this ? i.e : Is it possible to have baseUrl that will be prepended to the path of any further require or import calls ?
Basically if I do this : 
````
System.config({

	baseUrl : 'js/',
	
    paths: {
    	'*': '/js/*'
  	}
    // or 'traceur' or 'typescript'
    transpiler: 'babel',
    // or traceurOptions or typescriptOptions
    babelOptions: {

    }
});


var app = require("App.js");

console.log(app)
````

Neither of the  ````paths````` or the ````baseUrl```` option works :/
Thank you !
Alvin

**Comment by guybedford** [comment]
@alvinsight baseURL and paths are still supported fine. Are you running in the browser? The above should result in an import to `/js/App.js`? What are you getting?

**Comment by alvinsight** [comment]
It doesn't look like either is working for me :/
Yes I'm running it in the browser
I get a 404 because it's looking for App.js at the root : 
````
GET http://localhost:8080/systemJs/App.js 404 (Not Found)
````

**Comment by guybedford** [comment]
@alvinsight can you share your values from the console for `System.baseURL`, `System.paths['*']` and `document.baseURI`?

**Comment by alvinsight** [comment]
Sure, so I've amended the code snippet above to show you where I put the logs in case if might affect anything, and the results are as follows : 
````
System.baseURL :  http://localhost:8080/systemJs/
System.paths['*'] :  /js/*
document.baseURI :  http://localhost:8080/systemJs/
````

**Comment by guybedford** [comment]
@alvinsight it's all ok actually - you just need to call `System.import('App.js').then(console.log.bind(console))` rather here.

**Comment by guybedford** [comment]
Note also it is `System.config({ baseURL: '...' })` case-sensitive uppercase URL.

**Comment by alvinsight** [comment]
All right,  thanks, so the : ````var app = require("App.js");```` was never meant to work ?

**Comment by guybedford** [comment]
`require` is synchronous and means that the module is CommonJS. The way SystemJS, as an async loader, handles this is by loading and resolving the module before executing any code. That is, your config calls run after the module has already been resolved.

**Comment by alvinsight** [comment]
@guybedford Right, thanks, are you on gmail or gitChat ? It might be easier to send you a few question there rather than polluting this issue :)

**Comment by guybedford** [comment]
You can join gitter to ask questions. Do read the docs carefully though as this could have been avoided.
